lib/controller/controller.go

```
@@ -568,130 +640,181 @@ func (ctrl *ProvisionController) Run(stopCh <-chan struct{}) {
-	ctrl.failedProvisionThreshold = threshold
-	ctrl.failedProvisionStatsMutex.Unlock()
+func (ctrl *ProvisionController) runVolumeWorker() {
+	for ctrl.processNextVolumeWorkItem() {
+	}
 }
 
-// SetFailedDeleteThreshold sets the value of failedDeleteThreshold
-func (ctrl *ProvisionController) SetFailedDeleteThreshold(threshold int) {
-	ctrl.failedDeleteStatsMutex.Lock()
-	ctrl.failedDeleteThreshold = threshold
-	ctrl.failedDeleteStatsMutex.Unlock()
-}
+// processNextClaimWorkItem processes items from claimQueue
+func (ctrl *ProvisionController) processNextClaimWorkItem() bool {
+	obj, shutdown := ctrl.claimQueue.Get()
 
-// On add claim, check if the added claim should have a volume provisioned for
-// it and provision one if so.
-func (ctrl *ProvisionController) addClaim(obj interface{}) {
-	claim, ok := obj.(*v1.PersistentVolumeClaim)
-	if !ok {
-		glog.Errorf("Expected PersistentVolumeClaim but addClaim received %+v", obj)
-		return
+	if shutdown {
+		return false
 	}
 
-	if ctrl.shouldProvision(claim) {
-		ctrl.leaderElectorsMutex.Lock()
-		le, ok := ctrl.leaderElectors[claim.UID]
-		ctrl.leaderElectorsMutex.Unlock()
-		if ok && le.IsLeader() {
-			opName := fmt.Sprintf("provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-			ctrl.scheduleOperation(opName, func() error {
-				startTime := time.Now()
-				err := ctrl.provisionClaimOperation(claim)
-				ctrl.updateProvisionStats(claim, err, startTime)
-				return err
-			})
-		} else {
-			opName := fmt.Sprintf("lock-provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-			ctrl.scheduleOperation(opName, func() error {
-				ctrl.lockProvisionClaimOperation(claim)
-				return nil
-			})
+	err := func(obj interface{}) error {
+		defer ctrl.claimQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			ctrl.claimQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
```

probably we can simply return this error here, because outside function will call `utilruntime.HandleError` if err != nil.

---

done